// git-parse/dist/parse_commit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _git_log_format_markers = require("./constants/git_log_format_markers");

const parseCommit = commit => {
  const hash = commit[0];
  const authorName = commit[1];
  const authorEmail = commit[2];
  const date = commit[3];
  const messageIndex = commit.findIndex(line => line.match(_git_log_format_markers.gitLogMessageMarker));
  const fileIndex = commit.findIndex(line => line.match(_git_log_format_markers.gitLogFileMarker));
  const message = commit.slice(messageIndex + 1, fileIndex).join('\n');
  const files = commit.slice(fileIndex + 1);
  const addPattern = /^A\s([^\s]+)/;
  const deletePattern = /^D\s([^\s]+)/;
  const modifyPattern = /^M\s([^\s]+)/;
  const renamePattern = /^R[0-9]+\s(.+)\s(.+)/;

  const filterFileChanges = pattern => {
    return files.reduce((accumulator, file) => {
      const match = file.match(pattern);

      if (match) {
        accumulator.push({
          path: match[1]
        });
      }

      return accumulator;
    }, []);
  };

  const filesRenamed = files.reduce((accumulator, file) => {
    const match = file.match(renamePattern);

    if (match) {
      accumulator.push({
        oldPath: match[1],
        newPath: match[2]
      });
    }

    return accumulator;
  }, []);
  const parsedCommit = {
    hash,
    authorName,
    authorEmail,
    date,
    message,
    filesAdded: filterFileChanges(addPattern),
    filesDeleted: filterFileChanges(deletePattern),
    filesModified: filterFileChanges(modifyPattern),
    filesRenamed
  };
  return parsedCommit;
};

var _default = parseCommit;
exports.default = _default;
